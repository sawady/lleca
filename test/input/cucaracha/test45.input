
fun div(x : Int, y : Int) : Int {
    r := 0
    while x >= y {
        x := x - y
        r := r + 1
    }
    return r
}

fun mod(x : Int, y : Int) : Int {
    while x >= y {
        x := x - y
    }
    return x
}

fun testDivMod(x : Int, y : Int) {
  putNum(x)
  putChar(32)
  putChar(100)
  putChar(105)
  putChar(118)
  putChar(32)
  putNum(y)
  putChar(32)
  putChar(61)
  putChar(32)
  putNum(div(x, y))
  putChar(32)
  putChar(124)
  putChar(32)
  putNum(x)
  putChar(32)
  putChar(109)
  putChar(111)
  putChar(100)
  putChar(32)
  putNum(y)
  putChar(32)
  putChar(61)
  putChar(32)
  putNum(mod(x, y))
  putChar(10)
}

fun testDivModTo(x : Int, y : Int, n : Int) {
    while x < n {
        testDivMod(x, y)
        x := x + 1
    }
}

fun main() {
    testDivModTo(0, 2, 10)
    testDivModTo(0, 3, 10)
    testDivModTo(0, 7, 100)
    testDivModTo(0, 10, 100)
}

