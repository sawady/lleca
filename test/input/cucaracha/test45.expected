Cons(
  Function(
    div,
    Cons(
      Parameter(
        x,
        Int
      ),
      Cons(
        Parameter(
          y,
          Int
        ),
        Nil
      )
    ),
    Int,
    Cons(
      StmtAssign(
        r,
        ExprConstNum(
          0
        )
      ),
      Cons(
        StmtWhile(
          ExprGe(
            ExprVar(
              x
            ),
            ExprVar(
              y
            )
          ),
          Cons(
            StmtAssign(
              x,
              ExprSub(
                ExprVar(
                  x
                ),
                ExprVar(
                  y
                )
              )
            ),
            Cons(
              StmtAssign(
                r,
                ExprAdd(
                  ExprVar(
                    r
                  ),
                  ExprConstNum(
                    1
                  )
                )
              ),
              Nil
            )
          )
        ),
        Cons(
          StmtReturn(
            ExprVar(
              r
            )
          ),
          Nil
        )
      )
    )
  ),
  Cons(
    Function(
      mod,
      Cons(
        Parameter(
          x,
          Int
        ),
        Cons(
          Parameter(
            y,
            Int
          ),
          Nil
        )
      ),
      Int,
      Cons(
        StmtWhile(
          ExprGe(
            ExprVar(
              x
            ),
            ExprVar(
              y
            )
          ),
          Cons(
            StmtAssign(
              x,
              ExprSub(
                ExprVar(
                  x
                ),
                ExprVar(
                  y
                )
              )
            ),
            Nil
          )
        ),
        Cons(
          StmtReturn(
            ExprVar(
              x
            )
          ),
          Nil
        )
      )
    ),
    Cons(
      Function(
        testDivMod,
        Cons(
          Parameter(
            x,
            Int
          ),
          Cons(
            Parameter(
              y,
              Int
            ),
            Nil
          )
        ),
        Unit,
        Cons(
          StmtCall(
            putNum,
            Cons(
              ExprVar(
                x
              ),
              Nil
            )
          ),
          Cons(
            StmtCall(
              putChar,
              Cons(
                ExprConstNum(
                  32
                ),
                Nil
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    100
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      105
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        118
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      putChar,
                      Cons(
                        ExprConstNum(
                          32
                        ),
                        Nil
                      )
                    ),
                    Cons(
                      StmtCall(
                        putNum,
                        Cons(
                          ExprVar(
                            y
                          ),
                          Nil
                        )
                      ),
                      Cons(
                        StmtCall(
                          putChar,
                          Cons(
                            ExprConstNum(
                              32
                            ),
                            Nil
                          )
                        ),
                        Cons(
                          StmtCall(
                            putChar,
                            Cons(
                              ExprConstNum(
                                61
                              ),
                              Nil
                            )
                          ),
                          Cons(
                            StmtCall(
                              putChar,
                              Cons(
                                ExprConstNum(
                                  32
                                ),
                                Nil
                              )
                            ),
                            Cons(
                              StmtCall(
                                putNum,
                                Cons(
                                  ExprCall(
                                    div,
                                    Cons(
                                      ExprVar(
                                        x
                                      ),
                                      Cons(
                                        ExprVar(
                                          y
                                        ),
                                        Nil
                                      )
                                    )
                                  ),
                                  Nil
                                )
                              ),
                              Cons(
                                StmtCall(
                                  putChar,
                                  Cons(
                                    ExprConstNum(
                                      32
                                    ),
                                    Nil
                                  )
                                ),
                                Cons(
                                  StmtCall(
                                    putChar,
                                    Cons(
                                      ExprConstNum(
                                        124
                                      ),
                                      Nil
                                    )
                                  ),
                                  Cons(
                                    StmtCall(
                                      putChar,
                                      Cons(
                                        ExprConstNum(
                                          32
                                        ),
                                        Nil
                                      )
                                    ),
                                    Cons(
                                      StmtCall(
                                        putNum,
                                        Cons(
                                          ExprVar(
                                            x
                                          ),
                                          Nil
                                        )
                                      ),
                                      Cons(
                                        StmtCall(
                                          putChar,
                                          Cons(
                                            ExprConstNum(
                                              32
                                            ),
                                            Nil
                                          )
                                        ),
                                        Cons(
                                          StmtCall(
                                            putChar,
                                            Cons(
                                              ExprConstNum(
                                                109
                                              ),
                                              Nil
                                            )
                                          ),
                                          Cons(
                                            StmtCall(
                                              putChar,
                                              Cons(
                                                ExprConstNum(
                                                  111
                                                ),
                                                Nil
                                              )
                                            ),
                                            Cons(
                                              StmtCall(
                                                putChar,
                                                Cons(
                                                  ExprConstNum(
                                                    100
                                                  ),
                                                  Nil
                                                )
                                              ),
                                              Cons(
                                                StmtCall(
                                                  putChar,
                                                  Cons(
                                                    ExprConstNum(
                                                      32
                                                    ),
                                                    Nil
                                                  )
                                                ),
                                                Cons(
                                                  StmtCall(
                                                    putNum,
                                                    Cons(
                                                      ExprVar(
                                                        y
                                                      ),
                                                      Nil
                                                    )
                                                  ),
                                                  Cons(
                                                    StmtCall(
                                                      putChar,
                                                      Cons(
                                                        ExprConstNum(
                                                          32
                                                        ),
                                                        Nil
                                                      )
                                                    ),
                                                    Cons(
                                                      StmtCall(
                                                        putChar,
                                                        Cons(
                                                          ExprConstNum(
                                                            61
                                                          ),
                                                          Nil
                                                        )
                                                      ),
                                                      Cons(
                                                        StmtCall(
                                                          putChar,
                                                          Cons(
                                                            ExprConstNum(
                                                              32
                                                            ),
                                                            Nil
                                                          )
                                                        ),
                                                        Cons(
                                                          StmtCall(
                                                            putNum,
                                                            Cons(
                                                              ExprCall(
                                                                mod,
                                                                Cons(
                                                                  ExprVar(
                                                                    x
                                                                  ),
                                                                  Cons(
                                                                    ExprVar(
                                                                      y
                                                                    ),
                                                                    Nil
                                                                  )
                                                                )
                                                              ),
                                                              Nil
                                                            )
                                                          ),
                                                          Cons(
                                                            StmtCall(
                                                              putChar,
                                                              Cons(
                                                                ExprConstNum(
                                                                  10
                                                                ),
                                                                Nil
                                                              )
                                                            ),
                                                            Nil
                                                          )
                                                        )
                                                      )
                                                    )
                                                  )
                                                )
                                              )
                                            )
                                          )
                                        )
                                      )
                                    )
                                  )
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        )
      ),
      Cons(
        Function(
          testDivModTo,
          Cons(
            Parameter(
              x,
              Int
            ),
            Cons(
              Parameter(
                y,
                Int
              ),
              Cons(
                Parameter(
                  n,
                  Int
                ),
                Nil
              )
            )
          ),
          Unit,
          Cons(
            StmtWhile(
              ExprLt(
                ExprVar(
                  x
                ),
                ExprVar(
                  n
                )
              ),
              Cons(
                StmtCall(
                  testDivMod,
                  Cons(
                    ExprVar(
                      x
                    ),
                    Cons(
                      ExprVar(
                        y
                      ),
                      Nil
                    )
                  )
                ),
                Cons(
                  StmtAssign(
                    x,
                    ExprAdd(
                      ExprVar(
                        x
                      ),
                      ExprConstNum(
                        1
                      )
                    )
                  ),
                  Nil
                )
              )
            ),
            Nil
          )
        ),
        Cons(
          Function(
            main,
            Nil,
            Unit,
            Cons(
              StmtCall(
                testDivModTo,
                Cons(
                  ExprConstNum(
                    0
                  ),
                  Cons(
                    ExprConstNum(
                      2
                    ),
                    Cons(
                      ExprConstNum(
                        10
                      ),
                      Nil
                    )
                  )
                )
              ),
              Cons(
                StmtCall(
                  testDivModTo,
                  Cons(
                    ExprConstNum(
                      0
                    ),
                    Cons(
                      ExprConstNum(
                        3
                      ),
                      Cons(
                        ExprConstNum(
                          10
                        ),
                        Nil
                      )
                    )
                  )
                ),
                Cons(
                  StmtCall(
                    testDivModTo,
                    Cons(
                      ExprConstNum(
                        0
                      ),
                      Cons(
                        ExprConstNum(
                          7
                        ),
                        Cons(
                          ExprConstNum(
                            100
                          ),
                          Nil
                        )
                      )
                    )
                  ),
                  Cons(
                    StmtCall(
                      testDivModTo,
                      Cons(
                        ExprConstNum(
                          0
                        ),
                        Cons(
                          ExprConstNum(
                            10
                          ),
                          Cons(
                            ExprConstNum(
                              100
                            ),
                            Nil
                          )
                        )
                      )
                    ),
                    Nil
                  )
                )
              )
            )
          ),
          Nil
        )
      )
    )
  )
)
